### ISSUE TITLE
Abiity to write long lines

### ISSUE THREAD

**Issue by certik** [lfortran/fortran_proposals#9]
Ability to write arbitrary long lines from files to avoid MAX_LINE_LENGTH etc. Should be possible with allocatable character variable. Asked by [Pierre de Buyl](https://twitter.com/pdebuyl/status/1184427114185015297).

**Comment by pdebuyl** [comment]
In this example program, I set an arbitrary value `MAX_LINE_LENGTH`.
```
program read_lines
  implicit none

  type string_t
     character(len=:), allocatable :: s
  end type string_t

  integer, parameter :: MAX_LINE_LENGTH=32
  integer :: i, u, n_lines
  character(len=MAX_LINE_LENGTH) :: current_line
  type(string_t), allocatable :: lines(:)

  write(*,*) 'Number of lines to read?'
  read(*,*) n_lines

  open(newunit=u, file='lines.txt')

  allocate(lines(n_lines))

  do i = 1, n_lines
     read(u, '(a)') current_line
     lines(i)%s = trim(current_line)
  end do

  do i = 1, n_lines
     write(*,'(a)') lines(i)%s
  end do

end program read_lines
```

Ideally, I could replace this with an "allocation on `read`". The syntax below is just an example, of course.

```
program read_lines
  implicit none

  type string_t
     character(len=:), allocatable :: s
  end type string_t

  integer :: i, u, n_lines
  type(string_t), allocatable :: lines(:)

  write(*,*) 'Number of lines to read?'
  read(*,*) n_lines

  open(newunit=u, file='lines.txt')

  allocate(lines(n_lines))

  do i = 1, n_lines
     read(u, '(a)', alloc_on_read=.true.) lines(i)%s
  end do

  do i = 1, n_lines
     write(*,'(a)') lines(i)%s
  end do

end program read_lines
```

**Comment by certik** [comment]
@pdebuyl thank you. I think the proposal is for `read()` to automatically allocate your string variable to hold the full line, correct? I changed the title of the issue appropriately.

Let me discuss this with the members of the committee and see if this has a chance of getting in, and if so, write a paper for this.

**Comment by certik** [comment]
Here is a similar idea applied to a related problem: https://j3-fortran.org/doc/year/19/19-252.txt.

**Comment by pdebuyl** [comment]
@certik Yes, the usage I have in mind is specifically for read statements.

In "J3/19-252", the automatic allocation is planned for error messages and function return values. I think that my proposal would come in addition to it.

**Comment by certik** [comment]
@pdebuyl yes, that would be in addition to it. I posted it as an example that your proposal is not something unprecedented.

**Comment by ivan-pi** [comment]
VAX Fortran had an additional 'Q' descriptor that caused an integer variable to be assigned the length of the input line:
```
character*80 buffer
integer length

read(u,'(Q,A)') length, buffer
```
The purpose of this extension was to solve the problem of dealing with trailing blanks, i.e. to distinguish between trailing blanks of the empty buffer and those which are actually written in the file. The `length` variable would contain the length of what was actually read, discounting carriage control characters.

It would also be convenient to have this functionality for reading from standard input. The case of trailing blanks should be handled as well (see related question [here](https://stackoverflow.com/questions/50680306/how-can-i-know-the-length-of-user-input-with-trailing-whitespaces-in-fortran)).

**Comment by FortranFan** [comment]
@certik.

Fyi, you may have noticed the second-most popular item on the [**user feedback in 2017 to the WG5 survey**](https://isotc.iso.org/livelink/livelink?func=ll&objId=19530634&objAction=Open&viewType=1) for Fortran 2020 feature set was "Automatic Allocation on READ into ALLOCATABLE character or array".  That item looks very much similar to your proposal here. 

Some questions: the survey feedback item lists and "ALLOCATABLE character" or "ALLOCATABLE array".  Would you consider both for your proposal or just the character type?  If both, would you lean toward opening a separate issue for ALLOCATABLE array or modify this one along the lines of the survey item?

Thanks,

**Comment by certik** [comment]
@FortranFan thanks. For now I just put a comment in the issue description above. Once somebody starts drafting a proposal for this, we can decide if we should split it into two, or put everything into just one proposal.

**Comment by qolin1** [comment]
There is a current thread in comp.lang.fortran that discusses this, 

https://groups.google.com/forum/#!topic/comp.lang.fortran/9_0q0dBRzpk

The discussion starts off talking about stack sizes, but later (message of Ian Harvey, 9 jan) it verges into this proposal. Ian and Ron Shepard have an interesting exchange, wherein Ian expands on some of the problems it would cause. Alas I expect the Fortran standards committee will do the same, because AFAICS the problems are very real.

**Comment by urbanjost** [comment]
Are you looking to just be able to read a line of arbitrary length into a CHARACTER variable or looking to be able to read other arrays like an arbitrary number of numeric values?  Is this something like
     http://fortranwiki.org/fortran/show/readline

or something like the routines in
https://urbanjost.github.io/general-purpose-fortran/download/tmp/html/man3.html in the M_io module that read a file or numeric table into an array?  

I started looking at the discussion referenced above but it drifts thru quite a few topics.
